pkg/controller/atlascustomrole/atlascustomrole_controller.go

```diff
@@ -0,0 +1,245 @@
+// +kubebuilder:rbac:groups="",namespace=default,resources=events,verbs=create;patch
+
+func (r *AtlasCustomRoleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	atlasCustomRole := &akov2.AtlasCustomRole{}
+
+	err := r.Client.Get(ctx, req.NamespacedName, atlasCustomRole)
+	objectNotFound := err != nil && apiErrors.IsNotFound(err)
+	failedToRetrieve := err != nil && !objectNotFound
+
+	switch {
+	case failedToRetrieve:
+		return r.fail(req, err), nil
+	case objectNotFound:
+		return r.notFound(req), nil
+	}
+
+	if customresource.ReconciliationShouldBeSkipped(atlasCustomRole) {
+		return r.skip(), nil
+	}
+
+	r.Log.Infow("-> Starting AtlasCustomRole reconciliation", "spec", atlasCustomRole.Spec, "status",
+		atlasCustomRole.GetStatus())
+	conditions := akov2.InitCondition(atlasCustomRole, api.FalseCondition(api.ReadyType))
+	workflowCtx := workflow.NewContext(r.Log, conditions, ctx)
+	defer func() {
+		statushandler.Update(workflowCtx, r.Client, r.EventRecorder, atlasCustomRole)
+		r.Log.Infow("-> Finished AtlasCustomRole reconciliation", "spec", atlasCustomRole.Spec, "status",
+			atlasCustomRole.GetStatus())
+	}()
+
+	valid, err := customresource.ResourceVersionIsValid(atlasCustomRole)
+	if err != nil {
+		return r.terminate(workflowCtx, atlasCustomRole, api.ResourceVersionStatus, workflow.AtlasResourceVersionIsInvalid, true, err), nil
+	}
+
+	if !valid {
+		return r.terminate(workflowCtx,
+			atlasCustomRole,
+			api.ResourceVersionStatus,
+			workflow.AtlasResourceVersionMismatch,
+			true,
+			fmt.Errorf("version of the resource '%s' is higher than the operator version '%s'", atlasCustomRole.GetName(), version.Version)), nil
+	}
+	workflowCtx.SetConditionTrue(api.ResourceVersionStatus).SetConditionTrue(api.ValidationSucceeded)
+
+	if !r.AtlasProvider.IsResourceSupported(atlasCustomRole) {
+		return r.terminate(workflowCtx, atlasCustomRole,
+			api.ProjectCustomRolesReadyType, workflow.AtlasGovUnsupported,
+			false,
+			fmt.Errorf("the %T is not supported by Atlas for government", atlasCustomRole)), nil
+	}
+
+	credentials, err := selectCredentials(workflowCtx.Context, r.Client, atlasCustomRole)
+	if err != nil {
+		return r.fail(req, err), nil
+	}
+
+	atlasSdkClient, _, err := r.AtlasProvider.SdkClient(workflowCtx.Context, credentials, workflowCtx.Log)
+	if err != nil {
+		return r.terminate(workflowCtx,
+			atlasCustomRole,
+			api.ProjectCustomRolesReadyType,
+			workflow.AtlasAPIAccessNotConfigured,
+			true,
+			fmt.Errorf("unable to create atlas client: %s", err.Error())), nil
+	}
+	workflowCtx.SdkClient = atlasSdkClient
+	if res := handleCustomRole(workflowCtx, r.Client, atlasCustomRole, r.ObjectDeletionProtection); !res.IsOk() {
+		return r.fail(req, fmt.Errorf("%s", res.GetMessage())), nil
+	}
+	return r.idle(workflowCtx), nil
+}
+
+func selectCredentials(ctx context.Context, k8sClient client.Client, akoRole *akov2.AtlasCustomRole) (*client.ObjectKey, error) {
```

Comment:
One thing I like about this proposal is that it does not require any Atlas calls yet, only Kubernetes queries.

Maybe we could:
1) Get credentials first.
2) Populate the project ID using the credentials as an input, as they are needed when using Kubernetes references, and optionally to check the project actually exists with external references.

But I suggest we make this either common methods or functions we can reuse from anywhere, instead of repeating them on every new controller that uses the same references and credentials schema.